### PR TITLE
nixos/docker-containers: Rename to virtualisation.oci-containers.containers

### PR DESCRIPTION
**Pull request by adisbladis**
###### Motivation for this change
Using the `docker` daemon for the `docker-containers` never fit entirely in the NixOS model.
- There was no way to apply systemd limits to each container systemd service.
- It did not play nicely with systemd sandboxing features as the actual execution happens in the docker daemon context, not the systemd service.
- Setting users on the systemd service would have no real effect.

This PR implements a minimal set of changes to change the runtime. Any module improvements (like rootless declarative containers) we can make with `podman` should be left for a follow-up.

The potential downside is that users may be confused by the change when they run `docker ps` and the container is not showing up.

##### Discussion
After the discussion in this PR I have decided to change the scope of the PR: to only make the backend for this module configurable while not changing the default quite yet.
Both backends (`docker` & `podman`) have NixOS tests.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by saschagrunert — APPROVED** [review]
Nice! LGTM

**Comment by adisbladis** [comment]
@GrahamcOfBorg test docker-containers

**Review by vdemeester — APPROVED** [review]
This is nice :tada: 
LGTM :tiger:

**Comment by adisbladis** [comment]
@GrahamcOfBorg test docker-containers

**Comment by misuzu** [comment]
Maybe this should be renamed `docker-containers` -> `podman-containers` since this is not really docker containers anymore?

File: nixos/modules/virtualisation/docker-containers.nix
Comment (misuzu) [inline nixos/modules/virtualisation/docker-containers.nix:212]: These lines doesn't make sense, it's better to drop them.
```
    after = mkAfter;
    requires = after;
```

File: nixos/modules/virtualisation/docker-containers.nix
Comment (misuzu) [inline nixos/modules/virtualisation/docker-containers.nix:61]: How would this work? `podman` will only write to `journald` or to `stdout` too?

**Comment by adisbladis** [comment]
> Maybe this should be renamed docker-containers -> podman-containers since this is not really docker containers anymore? And maybe put it under virtualisation namespace if we are at it.

I would rather not switch the name to `podman-containers` as I consider the runtime implementation to be unimportant.
I think `docker-containers` is a fine name even if we're not using `docker` as the underlying implementation any longer.

File: nixos/modules/virtualisation/docker-containers.nix
Comment (adisbladis) [reply nixos/modules/virtualisation/docker-containers.nix:212]: Sure it makes sense. These options are populated by the container `dependsOn` option.

Comment (misuzu) [reply nixos/modules/virtualisation/docker-containers.nix:212]: Yeah, `mkAfter` name is confusing.

Comment (adisbladis) [reply nixos/modules/virtualisation/docker-containers.nix:212]: I agree with that. I've renamed the variable to `dependsOn` and removed the recursive attrset.

**Comment by saschagrunert** [comment]
> Maybe this should be renamed `docker-containers` -> `podman-containers` since this is not really docker containers anymore? And maybe put it under `virtualisation` namespace if we are at it.

How about OCI containers?

File: nixos/modules/virtualisation/docker-containers.nix
Comment (adisbladis) [reply nixos/modules/virtualisation/docker-containers.nix:61]: `podman` will also write to stderr/stdout. 
I've updated the service definition with:
```
StandardOutput = "null";
StandardError = "null";
```

We'll leave logging entirely to `podman`.

Comment (adisbladis) [reply nixos/modules/virtualisation/docker-containers.nix:61]: Note: I'd kept the default "none" but podman doesn't have that one.

**Comment by adisbladis** [comment]
> How about OCI containers?

SGTM! I've renamed `docker-containers` to `virtualisation.oci-containers`.

**Comment by Profpatsch** [comment]
Now that’s what I call drop-in.

**Comment by danielfullmer** [comment]
Would this allow us to switch to using `cgroupsv2` in `systemd` by default? My understanding was that docker was holding us back for that change. Some previous discussion here: https://github.com/NixOS/nixpkgs/issues/73800 https://github.com/NixOS/nixpkgs/pull/68096

**Comment by saschagrunert** [comment]
For cgroupsv2 we have to wait for a new runc release or switch to crun.

**Comment by denibertovic** [comment]
As a NixOS user I'm super -1 on this. It's OK to add another module but removing docker-containers is going to suck for users like me who don't want to use podman (and the rest of that ecosystem). What's wrong with adding this as a separate module and leave the docker modules exist? I'm -1 on the name as well. Podman is not the only OCI compliant tool out there. It seems disingenuous to imply that.

**Comment by arianvp** [comment]
We should fix outstanding issues with podman before merging this.  

this one is pretty serious for example: https://github.com/NixOS/nixpkgs/issues/77925

**Comment by arianvp** [comment]
I also agree with @denibertovic  that naming it OCI is not fair.   How about just naming it `podman-containers` ? I also see no reason to drop the `docker-containers` module  for now; though we might want to throw an assertion error when enabling docker and cgroupsv2 at the same time in the future

**Comment by adisbladis** [comment]
> As a NixOS user I'm super -1 on this. It's OK to add another module but removing docker-containers is going to suck for users like me who don't want to use podman (and the rest of that ecosystem). 

Do you have any actual arguments for not using Podman & associated tooling or is this just FUD?

> What's wrong with adding this as a separate module and leave the docker modules exist?

Fragmentation and the fact that running Docker containers inside systemd units has surprising limitations like being unable to constrain a unit in various ways like `ReadWriteDirectories`, `CPUAccounting`, etc, etc.
This means the current Docker container module never really worked in the composable way you'd expect from NixOS.

I'm strongly against duplicating the current module under a new `podman-containers` module as this would be pretty much a carbon-copy of the current module.

What we could do quite easily is to provide a parameter which docker runtime you'd like to use, then anyone actually wanting to use the Docker daemon for whatever reason could do so.
This parameter should default to `podman` for above mentioned reasons.

> We should fix outstanding issues with podman before merging this.
> this one is pretty serious for example: #77925

I completely agree.
Maybe if we implement the runtime parameter that should be set to `docker` by default until #77925 is fixed.

**Comment by denibertovic** [comment]
> Do you have any actual arguments for not using Podman & associated tooling or is this just FUD?

@adisbladis  What kind of a response is that? Why would you imply malice straight away??!
I don't think I need to explain or list, potentially subjective, arguments why I want to use one tech over the other.
That would not be productive.
If this was a discussion about using apache instead of nignx (or the other way around I don't care) would I still need to present valid arguments why I want to be able to keep using my choice of program? I'm just saying
leave it to the user to choose whatever they want to use.

> Fragmentation and the fact that running Docker containers inside systemd units has surprising limitations like being unable to constrain a unit in various ways like ReadWriteDirectories, CPUAccounting, etc, etc.

Not being able to use `ReadWriteDirectories` and `CPUAccouting` is very far from `This means the current Docker container module never really worked in the composable way you'd expect from NixOS.`

I don't need those features for instance. I'm sure there are many users that use various tech in different ways than you and I. That should be encouraged and they should be able to choose that.

>I'm strongly against duplicating the current module under a new podman-containers module as this would be pretty much a carbon-copy of the current module.
>What we could do quite easily is to provide a parameter which docker runtime you'd like to use, then anyone actually wanting to use the Docker daemon for whatever reason could do so.
This parameter should default to podman for above mentioned reasons.

I don't have any particular opinion on how the implementation should go. I'm sure there are various abstractions that could be used to make this not be `a carbon copy`. My point was merely that we should allow users to be able run one or the other. 

I don't think there should be a default though. It should be clear to the user what is they're "activating". It would be very wrong, as someone pointed out above, if the user ran a container and then did "docker ps" and got zilch.
Nix already had enough bad UX to keep adding to it. Just my 2c.

**Comment by apeyroux** [comment]
> Do you have any actual arguments for not using Podman & associated tooling or is this just FUD?
> 
> >

My company uses docker and I want to be isofunctional with her for the dev.

**Comment by arianvp** [comment]
> I'm strongly against duplicating the current module under a new podman-containers module as this would be pretty much a carbon-copy of the current module

Luckily nix is a language and we  can abstract these things no?

I really would like to avoid breaking compatibility for the sake of breaking it. If this PR is merged I would imagine something like this:

Oci-container module has indeed some kind of backend parameter that's either podman, docker or others.

There is a mkRenameModuleOption or equivalent that translates the old `docker-containers` module option to the new one and emits an optional warning about deprecation.

Renaming a module without any form of depreciation or `mkRenameModule`  sounds like a bad idea to me as it just breaks people's existing configs without any good reason. And also doesn't guide the user to do the right thing when `nixos-rebuild switch`'ing

**Comment by arianvp** [comment]
Concretely:

I would suggest adding this to the module:

1. add an option called `virtualisation.oci-containers.backend` which can either be `docker` or `podman`

2. add the following `mkChangedOptionModule` so that existing `docker-containers` modules get automatically translated to `virtualisaiton.oci-containers` whilst issueing a warning to the user

```nix
    (lib.mkChangedOptionModule
      [ "docker-containers"  ]
      [ "virtualisation" "oci-containers" ]
      (oldcfg: lib.mkMerge [
        oldcfg 
        {backend = "docker";}
      ]))
```

This seems to both make @adisbladis happy as we dont have to maintain two modules and @denibertovic happy as they can keep using docker and get a hint on how to migrate.

**Comment by adisbladis** [comment]
According to the suggestion by @arianvp I've gone ahead and change the purpose of this PR.
Until https://github.com/NixOS/nixpkgs/issues/77925 is resolved we cannot merge the change of default runtime from `docker` to `podman`.

It's still my intent to do so but for now we're mainly changing the module name and making the backend configurable.

**Comment by adisbladis** [comment]
@GrahamcOfBorg test oci-containers

**Comment by adisbladis** [comment]
@GrahamcOfBorg test oci-containers

**Comment by arianvp** [comment]
Looks good to me now

**Comment by grahamc** [comment]
I just wrote up an issue about some issues with podman's overall packaging:  #86245

**Comment by adisbladis** [comment]
Unless I hear objections I plan to merge this PR within the next couple of days.

Note that the PR scope has changed, this is now only a module rename + it makes the backend configurable.
We are not changing any defaults (yet).

File: nixos/modules/virtualisation/oci-containers.nix
Comment (flokli) [inline nixos/modules/virtualisation/oci-containers.nix:227]: In the case of `podman`, do we need to set `environment = config.networking.proxy.envVars` here too?

For the docker backend, this unit only does IPC with the daemon (and the daemon has this already set properly), but for podman, it needs to inherit system-wide proxy servers IIRC.

Comment (adisbladis) [reply nixos/modules/virtualisation/oci-containers.nix:227]: Very good point. Maybe we should simply always set the proxy env vars regardless of backend?

Comment (flokli) [reply nixos/modules/virtualisation/oci-containers.nix:227]: SGTM.

**Comment by dasJ** [comment]
Apart from all the things this makes better, it also introduces an interesting side-effect.

Every time any part of the system path changes, `config.system.path` changes, which means that the `ExecStart` of the container units changes which causes the containers to be restarted on every configuration switch that also changes the system path in some way.

cc @ajs124

**Pull request merged**
